packages/cw721/src/state.rs

```diff
@@ -0,0 +1,460 @@
+    pub approvals: Vec<Approval>,
+
+    /// Universal resource identifier for this NFT
+    /// Should point to a JSON file that conforms to the ERC721
+    /// Metadata JSON Schema
+    pub token_uri: Option<String>,
+
+    /// You can add any custom metadata here when you extend cw721-base
+    pub extension: TNftMetadataExtension,
+}
+
+#[cw_serde]
+pub struct Approval {
+    /// Account that can transfer/send the token
+    pub spender: Addr,
+    /// When the Approval expires (maybe Expiration::never)
+    pub expires: Expiration,
+}
+
+impl Approval {
+    pub fn is_expired(&self, block: &BlockInfo) -> bool {
+        self.expires.is_expired(block)
+    }
+}
+
+pub struct TokenIndexes<'a, TNftMetadataExtension>
+where
+    TNftMetadataExtension: Cw721State,
+{
+    pub owner: MultiIndex<'a, Addr, NftInfo<TNftMetadataExtension>, String>,
+}
+
+impl<'a, TNftMetadataExtension> IndexList<NftInfo<TNftMetadataExtension>>
+    for TokenIndexes<'a, TNftMetadataExtension>
+where
+    TNftMetadataExtension: Cw721State,
+{
+    fn get_indexes(
+        &'_ self,
+    ) -> Box<dyn Iterator<Item = &'_ dyn Index<NftInfo<TNftMetadataExtension>>> + '_> {
+        let v: Vec<&dyn Index<NftInfo<TNftMetadataExtension>>> = vec![&self.owner];
+        Box::new(v.into_iter())
+    }
+}
+
+#[cw_serde]
+pub struct CollectionMetadata<TCollectionMetadataExtension> {
+    pub name: String,
+    pub symbol: String,
+    pub extension: TCollectionMetadataExtension,
+    pub updated_at: Timestamp,
+}
+
+impl<TCollectionMetadataExtension, TCollectionMetadataExtensionMsg>
+    StateFactory<CollectionMetadata<TCollectionMetadataExtension>>
+    for CollectionMetadataMsg<TCollectionMetadataExtensionMsg>
+where
+    TCollectionMetadataExtension: Cw721State,
+    TCollectionMetadataExtensionMsg: Cw721CustomMsg + StateFactory<TCollectionMetadataExtension>,
+{
+    fn create(
+        &self,
+        deps: Deps,
+        env: &cosmwasm_std::Env,
+        info: &cosmwasm_std::MessageInfo,
+        current: Option<&CollectionMetadata<TCollectionMetadataExtension>>,
+    ) -> Result<CollectionMetadata<TCollectionMetadataExtension>, Cw721ContractError> {
+        self.validate(deps, env, info, current)?;
+        match current {
+            // Some: update existing metadata
+            Some(current) => {
+                let mut updated = current.clone();
+                if let Some(name) = &self.name {
+                    updated.name = name.clone();
+                }
+                if let Some(symbol) = &self.symbol {
+                    updated.symbol = symbol.clone();
+                }
+                let current_extension = current.extension.clone();
+                let updated_extension =
+                    self.extension
+                        .create(deps, env, info, Some(&current_extension))?;
+                updated.extension = updated_extension;
+                Ok(updated)
```

Comment:
... and creates `CollectionMedataExtension`, including validation of optional props like description, royalties, start trading time, etc.